From: Tom Lendacky <thomas.lendacky@amd.com>
Date: Fri, 15 Dec 2017 09:16:54 -0800
Subject: x86/cpu/AMD: Add speculative control support for AMD
Patch-mainline: submitted on 2018/1/9
References: bnc#1068032 CVE-2017-5715

Add speculative control support for AMD processors. For AMD, speculative
control is indicated as follows:

  CPUID EAX=0x00000007, ECX=0x00 return EDX[26] indicates support for
  both IBRS and IBPB.

  CPUID EAX=0x80000008, ECX=0x00 return EBX[12] indicates support for
  just IBPB.

Signed-off-by: Tom Lendacky <thomas.lendacky@amd.com>
Signed-off-by: Tim Chen <tim.c.chen@linux.inte.com>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 arch/x86/include/asm/cpufeatures.h | 1 +
 arch/x86/kernel/cpu/scattered.c    | 5 +++++
 2 files changed, 6 insertions(+)

diff --git a/arch/x86/include/asm/cpufeatures.h b/arch/x86/include/asm/cpufeatures.h
index f00024529863..79b307b7c2db 100644
--- a/arch/x86/include/asm/cpufeatures.h
+++ b/arch/x86/include/asm/cpufeatures.h
@@ -272,6 +272,7 @@
 #define X86_FEATURE_CLZERO		(13*32+ 0) /* CLZERO instruction */
 #define X86_FEATURE_IRPERF		(13*32+ 1) /* Instructions Retired Count */
 #define X86_FEATURE_XSAVEERPTR		(13*32+ 2) /* Always save/restore FP error pointers */
+#define X86_FEATURE_IBPB		(13*32+12) /* Indirect Branch Prediction Barrier */
 
 /* Thermal and Power Management Leaf, CPUID level 0x00000006 (EAX), word 14 */
 #define X86_FEATURE_DTHERM		(14*32+ 0) /* Digital Thermal Sensor */
diff --git a/arch/x86/kernel/cpu/scattered.c b/arch/x86/kernel/cpu/scattered.c
index 8108005d618f..70b21710c10f 100644
--- a/arch/x86/kernel/cpu/scattered.c
+++ b/arch/x86/kernel/cpu/scattered.c
@@ -76,6 +76,11 @@ void init_scattered_cpuid_features(struct cpuinfo_x86 *c)
 				sysctl_ibrs_enabled = 1;
 			if (ibpb_inuse)
 				sysctl_ibpb_enabled = 1;
+		} else if (boot_cpu_has(X86_FEATURE_IBPB)) {
+			printk_once(KERN_INFO "FEATURE IBPB Present\n");
+			set_ibpb_supported();
+			if (ibpb_inuse)
+				sysctl_ibpb_enabled = 1;
 		} else {
 			printk(KERN_INFO "FEATURE SPEC_CTRL Not Present\n");
 		}
-- 
2.15.1

